### DISCUSSION TITLE
Deteksi Anomali

### DISCUSSION THREAD

**Discussion by joejnt** [teamalgoritma/community#826] [Q&A] [answered]
Dear Tim Algoritma,

Mohon bantuannya, saya melakukan deteksi anomali menggunakan packages anomalize. Parameter yang saya gunakan adalah seperti berikut:

alpha_vol = 0.025
max_an_vol = 0.2

anomaly3 <- tib_trx_3 %>% 
  time_decompose(
    AMOUNT, method = "stl", frequency = "1 hour", trend = "12 hours",merge = T
    ) %>% 
  anomalize(
    remainder,
    method = "iqr",
    alpha =alpha_vol,
    max_anoms = max_an_vol) %>% 
  time_recompose()

Hasilnya masih banyak observasi yang terdeteksi sebagai anomali. dilihat dari plot seperti berikut:
![image](https://github.com/teamalgoritma/community/assets/103427989/160eec63-b306-4b90-804c-a09ea1057dfa)

Untuk dataset yang saya gunakan, adalah data 1 hari dan tidak saya lakukan agregasi. Pengaturan alpha sudah saya longgarkan, namun tetap masih banyak yang terdeteksi sebagai anomali. Apakah ada cara untuk mencari nilai yang terbaik dari alpha sama max_anoms, begitu juga untuk dekomposisi waktunya apakah ada cara untuk mencari frekuensi dan trend yang terbaik supaya mengurangi jumlah observasi yang terdeteksi sebagai anomali.

Tambahan: bagaimana ya cara menjelaskan data tersebut terdeteksi sebagai anomali klo dilihat dari nilai observerd, trend, remainder dan data aslinya. 

Terima kasih,

Best regards,
JT

**Comment by Yusraf002** [comment]
Dear @joejnt 

Pertanyaan Anda telah kami terima dan sedang kami usahakan terkait solusinya. Mohon kesediaannya untuk menunggu, terima kasih.

Best regards,
Team Algoritma

**Comment by ramdhinafinita** [comment, marked as answer]
Dear @joejnt 

Dari plot anomali yang Anda berikan, terlihat bahwa banyak observasi terdeteksi sebagai anomali, terutama pada periode malam hari. Salah satu kemungkinan penyebab terdeteksinya banyak anomali adalah kurang tepatnya dalam menentukan frekuensi dan trend. Hal ini bisa disebabkan oleh beberapa faktor:

**1. Frekuensi dan Tren:**

- **Frekuensi 1 jam:** Frekuensi 1 jam mungkin terlalu tinggi untuk data yang memiliki tren musiman 12 jam. Hal ini bisa menyebabkan algoritma deteksi anomali mendeteksi fluktuasi normal sebagai anomali, terutama pada periode malam hari di mana aktivitas umumnya lebih rendah.
- **Trend 12 jam:** Tren 12 jam mungkin tidak cukup untuk menangkap pola musiman data Anda secara akurat. Pertimbangkan menggunakan tren yang lebih kompleks (misalnya tren musiman dengan periode 24 jam) untuk lebih mencerminkan pola data Anda.

**2. Algoritma Deteksi Anomali:**

- **IQR:** Algoritma IQR menggunakan median dan kuartil untuk menentukan batas anomali. Pada dataset dengan distribusi data yang tidak normal, IQR bisa menghasilkan deteksi anomali yang kurang akurat. Pertimbangkan metode deteksi anomali lain (misalnya z-score, Isolation Forest) yang lebih robust terhadap distribusi data.


**Saran:**

- **Uji coba:** 
1. Pola data: Jika data Anda memiliki pola musiman yang lebih kompleks atau tren yang tidak linear, frekuensi 1 jam dan tren 12 jam mungkin tidak cukup untuk menangkapnya secara akurat.
2. Volume data: Pada dataset dengan volume data tinggi, fluktuasi natural data dapat terdeteksi sebagai anomali. Agregasi data (misalnya per 2 jam, per 4 jam) dapat membantu meminimalisir fluktuasi acak.


**Berikut beberapa tips untuk meningkatkan akurasi deteksi anomali:**
- **Gunakan frekuensi yang lebih rendah:** Cobalah frekuensi 4 jam atau 12 jam untuk menangkap tren musiman data Anda dengan lebih baik.
- **Gunakan tren yang lebih kompleks:** Pertimbangkan tren musiman dengan periode 24 jam atau tren non-linear untuk lebih mencerminkan pola data Anda.
- **Uji coba metode deteksi anomali lain:** Cobalah z-score atau Isolation Forest untuk mendapatkan hasil yang lebih robust terhadap distribusi data.
- **Agregasi data:** Pertimbangkan agregasi data per jam atau per hari untuk meringkas informasi dan meminimalisir fluktuasi acak.

**Reply by joejnt** [reply]
Dear @ramdhinafinita 

Terima kasih untuk insidenya, saya masih kurang paham pada poin:

**Gunakan frekuensi yang lebih rendah: Cobalah frekuensi 4 jam atau 12 jam untuk menangkap tren musiman data Anda dengan lebih baik.
Gunakan tren yang lebih kompleks: Pertimbangkan tren musiman dengan periode 24 jam atau tren non-linear untuk lebih mencerminkan pola data Anda.**

Mohon bisa dielaborasi lagi.

Best regards,
JT

**Reply by ramdhinafinita** [reply]
Secara umum maksudnya adalah pola trend yang dibentuk bisa dicoba digeneralisasi lagi per berapa jam, tidak terlalu tinggi di 1 jam.
Namun, bisa juga coba di ubah pada max anomali nya karena jika kita stel 0,2 itu terlalu besar, bisa coba dikecilkan lagi 5 % anomali atau 0.05.

Berikut ini bisa sebagai referensi untuk uji coba: 
https://cran.r-project.org/web/packages/anomalize/vignettes/anomalize_quick_start_guide.html

**Reply by joejnt** [reply]
@ramdhinafinita 

Well noted, terima kasih untuk insidenya.

Best regards
JT
